spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/annotation/SleuthAdvisorConfig.java

```diff
@@ -183,44 +183,138 @@ public Object invoke(MethodInvocation invocation) throws Throwable {
 		if (method == null) {
 			return invocation.proceed();
 		}
+
 		Method mostSpecificMethod = AopUtils
 				.getMostSpecificMethod(method, invocation.getThis().getClass());
 		NewSpan newSpan = SleuthAnnotationUtils.findAnnotation(mostSpecificMethod, NewSpan.class);
 		ContinueSpan continueSpan = SleuthAnnotationUtils.findAnnotation(mostSpecificMethod, ContinueSpan.class);
 		if (newSpan == null && continueSpan == null) {
 			return invocation.proceed();
 		}
+
+		if(isReactiveReturnType(method.getReturnType())){
+			return proceedUnderSpanReactively(invocation, newSpan, continueSpan);
+		} else {
+			return proceedUnderSpanSynchronously(invocation, newSpan, continueSpan);
+		}
+	}
+
+	private boolean isReactiveReturnType(Class<?> returnType) {
+		return Flux.class.equals(returnType) || Mono.class.equals(returnType);
+	}
+
+	private Object proceedUnderSpanSynchronously(
+			MethodInvocation invocation, NewSpan newSpan, ContinueSpan continueSpan) throws Throwable {
 		Span span = tracer().currentSpan();
 		if (newSpan != null || span == null) {
-			span = tracer().nextSpan().start();
+			span = tracer().nextSpan();
 			newSpanParser().parse(invocation, newSpan, span);
+			span.start();
 		}
 		String log = log(continueSpan);
 		boolean hasLog = StringUtils.hasText(log);
 		try (Tracer.SpanInScope ws = tracer().withSpanInScope(span)) {
-			if (hasLog) {
-				logEvent(span, log + ".before");
-			}
-			spanTagAnnotationHandler().addAnnotatedParameters(invocation);
-			addTags(invocation, span);
+			before(invocation, span, log, hasLog);
 			return invocation.proceed();
 		} catch (Exception e) {
-			if (logger.isDebugEnabled()) {
-				logger.debug("Exception occurred while trying to continue the pointcut", e);
-			}
-			if (hasLog) {
-				logEvent(span, log + ".afterFailure");
-			}
-			span.error(e);
+			onFailure(span, log, hasLog, e);
 			throw e;
 		} finally {
-			if (hasLog) {
-				logEvent(span, log + ".after");
+			after(span, newSpan != null, log, hasLog);
+		}
+	}
+
+	private Object proceedUnderSpanReactively(
+			MethodInvocation invocation, NewSpan newSpan, ContinueSpan continueSpan) throws Throwable{
+		boolean isNewSpan = newSpan != null;
+		Span spanPrevious = tracer().currentSpan();
+		Span span;
+		if (isNewSpan || spanPrevious == null) {
```

Comment:
make a comment mentioning how when spanPrevious was null, we somehow know to close it later

Reply:
Can you help me to clarify some points in original version of code?
In case of `@ContinueSpan` and span is not set before in tracer then new span will be created and not closed on method completion. 
In tests this case is not covered.
Is it by design?

Reply:
I think the case of continuespan we don't close. that's the main thing I think we should have a unit test about now. Correct me if I am wrong @marcingrzejszczak

Reply:
@adriancole @marcingrzejszczak please, check my last commit. I think it should fix issue with not closed spans on `@ContinueSpan`